### DISCUSSION TITLE
Integrating Jolt into a voxel world

### DISCUSSION THREAD

**Discussion by aecsocket** [jrouwe/JoltPhysics#446] [Q&A] [answered]
I'm writing an integration for Jolt into Minecraft, which has a voxel grid of blocks (these voxels can have varying box sizes and shapes, but they are always boxes and always fit within 1x(1-1.5)x1m). I'm wondering what the most efficient approach to integrating the two together would be.

My first attempt was a naive method of, on every game tick (1/20th sec) on the game thread, iterate through all active bodies, find nearby blocks, and add/remove static bodies from the engine based on those blocks. However this was very slow, and as I understand doesn't fully take advantage of Jolt as it means the static AABB has to always be rebuilt.

Is it maybe a better idea to generate an entire collision mesh (or set of bodies) when a single 16x256x16 chunk is loaded in the world? If so, would convex decomposition be useful here? Or maybe use a custom shape which looks up blocks from the world instead? How would this then work when a block in the chunk is modified?

**Comment by jrouwe** [comment, marked as answer]
The simplest approach would be indeed to divide your grid into chunks as you mention and then create a Body + StaticCompoundShape per chunk at load time of that chunk. You probably have a limited number of possible block sizes, so you could allocate just 1 BoxShape per block size and reuse it many times in the StaticCompoundShape. When the world changes because the player removes/places a block you would need to re-create the StaticCompoundShape for that chunk (which therefore should not be too big, 16x256x16 may already be too much, you could also go for e.g. 16x16x16). I don't see what convex decomposition would give you, your world already consists of convex shapes (boxes) so it is already implicitly decomposed.

A more advanced (and more efficient way) is to make a custom shape as you say (see [Create Custom Shapes on this page](https://jrouwe.github.io/JoltPhysics/). You would create a single body per world chunk and attach your custom shape to it (the Sub Shape IDs have only 32 bits so you can probably not encode the entire world with a single shape). You then need to implement the collision detection functions by checking if the query overlaps with a particular block and then forward the query to the BoxShape for that block. You can use the MutableCompoundShape as an example on how to check the bounding box of a query shape against the child shapes and forward the query to the child shape when it collides.

Note that when a block gets added/removed, you should query the space of the block (+ some tolerance around) for any sleeping bodies and wake them up. Jolt never wakes up bodies in these cases for performance reasons (typically streaming out large sections of data would cause everything to wake up and then those woken up objects would be removed immediately after, wasting precious CPU cycles).

**Reply by aecsocket** [reply]
Great, that's about what I guessed. Thanks for the info!

**Comment by aecsocket** [comment]
After implementing a basic solution using static compound shapes, where each 16x320x16 chunk is split into 20 vertical chunk slices, I've run into a problem. A minimum of 529 chunks are loaded, making 10,580 slices, each with up to 16^3 (= 4096) bodies per voxel inside them. This means that the RAM usage is insanely high, and ends up crashing the entire program.

Is there a better solution than generating a shape when each chunk is loaded? I'm thinking of detecting which chunk slices (i.e. 16x16x16 regions) all of the active bodies are interacting with, and generate just shapes for those instead. Could this be optimized by using the quadtrees (or some other implementation details) directly?

**Reply by jrouwe** [reply]
I think each slice should consist of a single **body** instead of 4096 bodies. Each body would have a single StaticCompoundShape with max 4096 shapes in them (where the shapes inside would be heavily reused because there's only a couple of variations). The overhead of adding a shape to a StaticCompoundShape is far less than adding a body, so I would expect this to use a lot less memory.

**Reply by aecsocket** [reply]
Thanks for the suggestion, this makes the memory usage a lot lower. Now the only thing I am concerned about is that there could potentially be many, many more chunks loaded at once than 529 (that's just the minimum), so I would still have to implement some mechanism to only generate collision bodies for *specific* chunk slices (i.e. slices which a body could interact with). To achieve this, what would be the best way to use Jolt's architecture to make this efficient? Broad phase query with an AABox?

**Reply by jrouwe** [reply]
That's a really hard question for me to answer. It really depends on the game engine architecture. E.g.:

* How many slices would you need max? 100K static bodies should not be an issue, so even if you have 5000 chunks loaded it should be fine. You can also play around with the chunk size (make it bigger) and find a decent trade off between memory and performance.
* Is there something in the game engine that stops simulating objects when they are too far away from the player(s)? In this case you could probably remove slices that are far away from the player(s) and regenerate them as player(s) gets closer.
* You could iterate over all active bodies and mark the slice (and neighboring slices depending on velocity) as active and make sure they're generated. If you go over a limit, remove least recently used slices.
* You can still go for the custom shape solution and have a body per slice / chunk but directly use the voxel data during collision queries as stated before. You could even create a simple custom DecoratedShape that on collision quickly generates a StaticCompoundShape and links it (with some logic to remove compounds that haven't been used in a while).

P.S. I'm not sure what you would gain with a broad phase query with an AABox. The voxel world is extremely regular, so if you generate a slice then you can store a reference to it in a 3D array and then it's trivial to determine if all slices close to the player / an active object have been generated.
